packages/@aws-cdk/aws-iot-actions/lib/s3-action.ts

```diff
@@ -0,0 +1,77 @@
+import * as iam from '@aws-cdk/aws-iam';
+import * as iot from '@aws-cdk/aws-iot';
+import * as s3 from '@aws-cdk/aws-s3';
+import { kebab as toKebabCase } from 'case';
+import { singletonActionRole } from './private/role';
+
+/**
+ * Configuration properties of an action for s3.
+ */
+export interface S3ActionProps {
+  /**
+   * The Amazon S3 canned ACL that controls access to the object identified by the object key.
+   * @see https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html#canned-acl
+   *
+   * @default None
+   */
+  readonly cannedAcl?: s3.BucketAccessControl;
+
+  /**
+   * The path to the file where the data is written.
+   *
+   * Supports substitution templates.
+   * @see https://docs.aws.amazon.com/iot/latest/developerguide/iot-substitution-templates.html
+   *
+   * @default '${topic()}/${timestamp()}'
```

Comment:
Should we have static, string-typed constants for these substitution templates? So that it's clear which ones can be used by which Actions?

Reply:
I think we cannot define all static, string-typed constants. Because users can use MQTT payload property defined by themselves in template.

For example, there are following MQTT payload:
```json
{
  "year": "2021",
  "month": "Nov",
  "day": "04",
  "data": 123.4
}
```

The user can define the bucket key as `${year}/${month}/${day}/${timestamp()}`.

Are we on the same page?

Reply:
I don't think we need to have all of them - but you have a few that are always available, right? Like `${timestamp()}` and `${topic()}`. Maybe it's worth it to include those that are always available, and just document the other ones?

Reply:
OK, I also want to make it as user-friendly as possible. So I'd like to be on same page!

1. there are [about 70 functions](https://docs.aws.amazon.com/iot/latest/developerguide/iot-sql-functions.html), and they take arguments
  - So we need to use functions, instead of static constants.
2. there are also [Operators](https://docs.aws.amazon.com/iot/latest/developerguide/iot-sql-operators.html).
  - So it is better that our functions return without `${}` as `topic()`.
3. in many cases, they are used embedded in strings without any regularity. For example, [cloudwatch metric name](https://docs.aws.amazon.com/iot/latest/developerguide/cloudwatch-metrics-rule-action.html#cloudwatch-metrics-rule-action-parameters) could be used substitution templates.

And I thought 3 plans.

### PLAN1, implement a set of functions

In light of the above, it might be better for us to implement a set of functions and for users to use them in template literals, as following:

```ts
key: `\${${iot.Template.topic(2)}}/\${${iot.Template.timestamp()}}`
```

### PLAN2, just validation

Or, is it better to implement validation as `JsonPath` of `@aws-cdk/aws-stepfunctions`
https://github.com/aws/aws-cdk/tree/master/packages/@aws-cdk/aws-stepfunctions/lib/fields.ts#L216-L226

For substitution templates, I think we will be using a lot of regular expressions.

### PLAN3, implement `Expression` and `Template`

Or, we implement `Expression` and `Template`?

I mean Expression as following:
```ts
iot.Expression.substring(iot.Expression.topic(2), 1)
// return type of Expression
```

I mean Template as following:
```ts
iot.Template.format(
  '{}/{}',
  iot.Expression.substring(iot.Expression.topic(2), 5, 10),
  iot.Expression.timestamp(),
)
// return type of Template
```

And, properties that can be used substitution templates will have the type `Template`.
```ts
{
  key: iot.Template,
}
```

The above may help to naturally communicate to users that substitution templates are available.

Reply:
Before I answer, I have one more clarifying question.

Are any of the substitutions Action-dependent? So, are there any substitutions that can only be used with certain Action(s)?

In any case, I see now this is a much more complicated topic than I initially thought, and should be tackled in a separate PR. Let's continue the discussion on how to best model them - that's fine, but I don't want to block this PR on the discussion. Let's proceed with this PR without anything around substitutions, like it is now.

Reply:
> Are any of the substitutions Action-dependent? So, are there any substitutions that can only be used with certain Action(s)?

As far as I can tell from the documentation, no. And from the documentation, they all seem to be common functions and not Action-dependent. 🙂

> Let's proceed with this PR without anything around substitutions, like it is now.

Sounds good!

Reply:
> > Are any of the substitutions Action-dependent? So, are there any substitutions that can only be used with certain Action(s)?
> 
> As far as I can tell from the documentation, no. And from the documentation, they all seem to be common functions and not Action-dependent. 🙂

Great! Given that, I would go with **Plan 1** from https://github.com/aws/aws-cdk/pull/17307#discussion_r743709980.